fix: filter None router logits in Qwen3 MoE and handle empty router logits (#39203)

SwiftAkira opened the pull request:
## What does this PR do?

This PR fixes issue #39203 where Qwen3 MoE models crash when mlp_only_layers is non-empty and output_router_logits=True. The issue occurs because MLP-only layers return None router logits, which are incorrectly collected and passed to load_balancing_loss_func, causing a TypeError.

## Root Cause Analysis

The problem was in the router logits collection logic in Qwen3MoeModel.forward(). Unlike Qwen2 MoE which properly filters None values, Qwen3 MoE was collecting all layer outputs without null checks:

- MLP-only layers (specified in mlp_only_layers) return None for router logits since they don't use expert routing
- The original code collected these None values into the router_logits tuple
- When load_balancing_loss_func processes this tuple, it fails on None entries

## Solution

This PR implements two complementary fixes:

1. **Router logits null check**: Added proper filtering during collection to match Qwen2 MoE pattern:
   ```python
   # Before (broken):
   if output_router_logits:
       all_router_logits += (layer_outputs[-1],)
   
   # After (fixed):
   if output_router_logits and layer_outputs[-1] is not None:
       all_router_logits += (layer_outputs[-1],)
   ```

2. **Empty tuple handling**: Added a custom load_balancing_loss_func that gracefully handles the edge case where all layers are MLP-only (resulting in an empty router_logits tuple):
   ```python
   if len(gate_logits) == 0:
       return 0
   ```

## Implementation Details

All changes were made in the modular architecture:
- **Source file**: src/transformers/models/qwen3_moe/modular_qwen3_moe.py (hand-edited)
- **Generated file**: src/transformers/models/qwen3_moe/modeling_qwen3_moe.py (auto-generated)

The fix follows the established pattern from Qwen2 MoE, ensuring consistency across the codebase.

## Testing

Comprehensive testing was performed with various configurations:

1. **Mixed configuration** (mlp_only_layers=[1,3]): 
   - Correctly collects 2 router logits from MoE layers
   - Successfully computes auxiliary loss
   
2. **All MoE configuration** (mlp_only_layers=[]):
   - Collects router logits from all layers
   - Standard auxiliary loss computation
   
3. **All MLP configuration** (mlp_only_layers=[0,1,2,3]):
   - Results in empty router logits tuple
   - Auxiliary loss returns 0 (no routing needed)

All test cases pass without errors.

## Backward Compatibility

This fix is fully backward compatible:
- Existing models continue to work unchanged
- Only adds null checks with minimal performance overhead
- Maintains the same API and behavior for valid configurations

## Fixes

Closes #39203

## How was this patch tested?

- Manual testing with Qwen3 MoE models using different mlp_only_layers configurations
- Verified proper router logits collection and auxiliary loss computation
- Tested edge cases including all-MLP and all-MoE scenarios
- Validated that no None values appear in the final router_logits tuple

cc @ArthurZucker @ntenenz

SwiftAkira committed:
fix: filter None router logits in Qwen3 MoE and handle empty router logits (#39203)

SwiftAkira committed:
style: format modular_qwen3_moe.py with ruff to pass CI

ArthurZucker reviewed (commented):
Hey #39120 was just merged, I would be more than happy if you can rebase and fix without having to add the filtering! SHould be straightforward

kaixuanliu committed:
fix bug using FSDP V1 will lead to model device not properly set (#39177)

* fix bug using FSDP V1 will lead to model device not properly set

Signed-off-by: Liu, Kaixuan <kaixuan.liu@intel.com>

* update the code

Signed-off-by: Liu, Kaixuan <kaixuan.liu@intel.com>

---------

Signed-off-by: Liu, Kaixuan <kaixuan.liu@intel.com>

xadupre committed:
Make _compute_dynamic_ntk_parameters exportable (#39171)

* Make _compute_dynamic_ntk_parameters exportable

* add unit test

Cyrilvallez committed:
[modular] Simplify logic and docstring handling (#39185)

* simplify a lot

* Update modular_model_converter.py

* finalize

* remove outdated functions

* apply it

* and examples

FightingZhen committed:
[bugfix] fix flash attention 2 unavailable error on Ascend NPU (#39166)

[bugfix] fix flash attention 2 error on Ascend NPU

ydshieh committed:
fix `fastspeech2_conformer` tests (#39229)

* fix

* fix

* fix

---------

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

qubvel committed:
RotaryEmbeddings change `is not None` -> `isinstance(..., dict)` (#39145)

is None -> isinstance dict

Cyrilvallez committed:
Fix patch helper (#39216)

remove -1

jiqing-feng committed:
enable xpu on kv-cache and hqq doc (#39246)

Signed-off-by: jiqing-feng <jiqing.feng@intel.com>

kaixuanliu committed:
adjust input and output texts for test_modeling_recurrent_gemma.py (#39190)

* adjust input and output texts for test_modeling_recurrent_gemma.py

Signed-off-by: Liu, Kaixuan <kaixuan.liu@intel.com>

* fix bug

Signed-off-by: Liu, Kaixuan <kaixuan.liu@intel.com>

* adjust

Signed-off-by: Liu, Kaixuan <kaixuan.liu@intel.com>

* update Expectation match

Signed-off-by: Liu, Kaixuan <kaixuan.liu@intel.com>

* fix

---------

Signed-off-by: Liu, Kaixuan <kaixuan.liu@intel.com>
Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

SwiftAkira committed:
fix: sync modular and main files kwargs signature for CI

SwiftAkira commented:
should work now just read above docu

SwiftAkira committed:
fix(qwen3_moe): resolve merge conflict

ArthurZucker reviewed (commented):
Sorry @SwiftAkira before I review could you fix your branch? Some rebasing seems to have gone wrong!

SwiftAkira committed:
fix: correct method signature for position_embeddings parameter

- Make position_embeddings an optional keyword argument instead of required positional
- Update gradient checkpointing calls to use keyword arguments
- Ensure backward compatibility with existing calling patterns
- Fix CI pipeline issues related to method signature mismatch

ArthurZucker commented on src/transformers/models/qwen3_moe/modeling_qwen3_moe.py:
```diff
@@ -489,26 +476,20 @@ def forward(
         output_hidden_states: Optional[bool] = None,
         output_router_logits: Optional[bool] = None,
```
```suggestion
```
as the `check_model_inputs` takes care of thiss

ArthurZucker commented on src/transformers/models/qwen3_moe/modeling_qwen3_moe.py:
```diff
@@ -308,16 +307,13 @@ def __init__(self, config: Qwen3MoeConfig, layer_idx: int):
-        use_cache: Optional[bool] = False,
         cache_position: Optional[torch.LongTensor] = None,
-        position_embeddings: Optional[tuple[torch.Tensor, torch.Tensor]] = None,  # necessary, but kept here for BC
         **kwargs: Unpack[FlashAttentionKwargs],
```
SHould be transformers kwargs!

ArthurZucker commented on src/transformers/models/qwen3_moe/modeling_qwen3_moe.py:
```diff
@@ -342,45 +338,36 @@ def forward(
-
-        if output_router_logits:
+        outputs = (hidden_states, self_attn_weights)
+        if router_logits is not None:
```
if you use the `check_model_inputs` decorator it will take care of this you only have to return the hidden states

ArthurZucker reviewed (commented):
History is still a bit messed up
